pkg/controller/cloneset/utils/pod_sorter.go

```diff
@@ -0,0 +1,298 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package utils
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"sort"
+	"strconv"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	podutil "k8s.io/kubernetes/pkg/api/v1/pod"
+	"k8s.io/utils/integer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type Ranker interface {
+	GetRank(*v1.Pod) float64
+}
+
+const (
+	// PodDeletionCost can be used to set to an int32 that represent the cost of deleting
+	// a pod compared to other pods belonging to the same ReplicaSet. Pods with lower
+	// deletion cost are preferred to be deleted before pods with higher deletion cost.
+	PodDeletionCost = "controller.kubernetes.io/pod-deletion-cost"
+)
+
+// ActivePodsWithRanks type allows custom sorting of pods so a controller can pick the best ones to delete.
+type ActivePodsWithRanks struct {
+	Pods          []*v1.Pod
+	Ranker        Ranker
+	AvailableFunc func(*v1.Pod) bool
+}
+
+func (s ActivePodsWithRanks) Len() int      { return len(s.Pods) }
+func (s ActivePodsWithRanks) Swap(i, j int) { s.Pods[i], s.Pods[j] = s.Pods[j], s.Pods[i] }
+
+func (s ActivePodsWithRanks) Less(i, j int) bool {
+	// 1. Unassigned < assigned
+	// If only one of the pods is unassigned, the unassigned one is smaller
+	if s.Pods[i].Spec.NodeName != s.Pods[j].Spec.NodeName && (len(s.Pods[i].Spec.NodeName) == 0 || len(s.Pods[j].Spec.NodeName) == 0) {
+		return len(s.Pods[i].Spec.NodeName) == 0
+	}
+	// 2. PodPending < PodUnknown < PodRunning
+	podPhaseToOrdinal := map[v1.PodPhase]int{v1.PodPending: 0, v1.PodUnknown: 1, v1.PodRunning: 2}
+	if podPhaseToOrdinal[s.Pods[i].Status.Phase] != podPhaseToOrdinal[s.Pods[j].Status.Phase] {
+		return podPhaseToOrdinal[s.Pods[i].Status.Phase] < podPhaseToOrdinal[s.Pods[j].Status.Phase]
+	}
+	// 3. Not available < available; Not ready < ready
+	// If only one of the pods is not ready, the not ready one is smaller
+	if s.AvailableFunc != nil {
+		if s.AvailableFunc(s.Pods[i]) != s.AvailableFunc(s.Pods[j]) {
+			return !s.AvailableFunc(s.Pods[i])
+		}
+	}
+	if podutil.IsPodReady(s.Pods[i]) != podutil.IsPodReady(s.Pods[j]) {
+		return !podutil.IsPodReady(s.Pods[i])
+	}
+
+	// 4. Lower pod-deletion cost < higher pod-deletion-cost
+	pi, _ := getDeletionCostFromPodAnnotations(s.Pods[i].Annotations)
+	pj, _ := getDeletionCostFromPodAnnotations(s.Pods[j].Annotations)
+	if pi != pj {
+		return pi < pj
+	}
+
+	// 5. Higher ranks < lower ranks
+	var rankI, rankJ float64
+	if s.Ranker != nil {
+		rankI = s.Ranker.GetRank(s.Pods[i])
+		rankJ = s.Ranker.GetRank(s.Pods[j])
+	}
+	if rankI != rankJ {
+		return rankI > rankJ
+	}
+
+	// TODO: take availability into account when we push minReadySeconds information from deployment into pods,
```

Comment:
is the TODO still valid here ?

Reply:
Yes, it comes from [k8s.io/kubernetes/pkg/controller](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/controller_utils.go#L852), and is still here yet.